File: apps/api/app/api/v1/country_pairs.py
from app.core.database import get_connection
from app.core.locales import LocaleQuery
from app.schemas.country_pairs import (
    CountryPairCompatibilityListResponse,
    CountryPairCompatibilityResponse,
)
from app.services import country_pairs as country_pairs_service
from fastapi import APIRouter, Depends
from psycopg import Connection
from typing import Annotated, Any


router = APIRouter(tags=["country-pairs"])


@router.get(
    "/country-pairs/{origin_slug}/{destination_slug}",
    response_model=CountryPairCompatibilityResponse,
)
def read_country_pair(
    origin_slug: str,
    destination_slug: str,
    connection: Annotated[Connection[Any], Depends(get_connection)],
    locale: LocaleQuery,
) -> CountryPairCompatibilityResponse:
    return country_pairs_service.get_country_pair_context(
        connection, origin_slug, destination_slug, locale
    )


@router.get(
    "/countries/{origin_slug}/destination-compatibility",
    response_model=CountryPairCompatibilityListResponse,
)
def read_destination_compatibility(
    origin_slug: str,
    connection: Annotated[Connection[Any], Depends(get_connection)],
    locale: LocaleQuery,
) -> CountryPairCompatibilityListResponse:
    return country_pairs_service.list_destination_pair_contexts(
        connection, origin_slug, locale
    )

File: apps/web/src/shared/api/country-pairs.ts
import type { components } from "@country-decision-atlas/contracts/generated/types";

import { DEFAULT_LOCALE } from "../lib/locale";
import { apiGet, queryString } from "./http";

type LocaleCode = components["schemas"]["LocaleCode"];
export type CountryPairCompatibilityResponse =
  components["schemas"]["CountryPairCompatibilityResponse"];
export type CountryPairCompatibilityListResponse =
  components["schemas"]["CountryPairCompatibilityListResponse"];

export function getCountryPair(
  originSlug: string,
  destinationSlug: string,
  locale: LocaleCode = DEFAULT_LOCALE,
): Promise<CountryPairCompatibilityResponse> {
  return apiGet<CountryPairCompatibilityResponse>(
    `/api/v1/country-pairs/${originSlug}/${destinationSlug}${queryString({ locale })}`,
  );
}

export function listDestinationCompatibility(
  originSlug: string,
  locale: LocaleCode = DEFAULT_LOCALE,
): Promise<CountryPairCompatibilityListResponse> {
  return apiGet<CountryPairCompatibilityListResponse>(
    `/api/v1/countries/${originSlug}/destination-compatibility${queryString({ locale })}`,
  );
}

export const countryPairsApi = {
  getCountryPair,
  listDestinationCompatibility,
};
